### PR TITLE
[WIP] Reworks wizard gamemode and the Lich spell

### LINKED ISSUE #8278: Bind Soul can be taken on Raging Mages, even though it doesn't work.
**Problem Description**:
You still get gibbed when you die as a wizard on raging mages, despite having an ensouled item

**What did you expect to happen**: 
To be able to revive using Bind Soul.

**What happened instead**:
Got gibbed on death, no chance to revive.

**Why is this bad/What are the consequences:**
It's 2 wasted points on an antag where every point matters. No indication of raging mages for the first mage, so you have no idea not to take it/do it.

**Steps to reproduce the problem**:
Die with bind soul active on raging mages. Become a pillar of salt.

### PR DESCRIPTION
**Pull request by datlo**
This PR refactors and reworks the wizard gamemode with the main objective being to have it last longer than 10 minutes and to give more people the opportunity to actually play wizard (and thus get better at it). It also reworks the Bind Soul (Lich) spell to make it work properly with respawning wizards.

Changes to the gamemode : 

- The wizard gamemode will now have a new wizard being spawned a short time after one dies. This will continue until the wizard cap is reached, by default 1 wizard per 25 player. Wizards are also considered dead if they are transformed into a non-human mob.

- Liches are only considered dead once they are permanently dead or transformed. This fixes #8278, liches will no longer gib on raging mages when they can revive.

- Raging mages had most of its procs and vars moved to the standard wizard gamemode but there should be no gameplay change to the gamemode other than the fix mentioned above.

Changes to summon guns and magic : 

- Summon guns and magic have been modified to delete the items summoned when the next wizard spawns. That way, later wizards are not impacted by the spell. 

- They are also both in the Challenge category, since thats the one they both belong to. The challenge category description was updated with a warning that new wizards should probably not make their life any harder than it already is.

Changes to Bind Soul (Lichdom)

- A lich's phylactery is now linked to the lich and will snap back to them if they are on a different zlevel.
- Liches are now dusted when their phylactery is destroyed.

TODO :

Fix bugs

🆑 
tweak: When a wizard die, a new wizard will spawn up to a cap scaling with server pop, 1 wizard per 25 players. Raging mages still work as it used to.
tweak; Summoned guns and magical items will disappear when another wizard is sent. They are now both in the Challenge spellbook category.
fix: Liches will no longer be gibbed on raging mages when they could revive.
tweak; A lich's phylactery will now snap back to the lich if they are on different zlevels.
tweak: Liches are now instantly dusted when their phylactery is destroyed.
/🆑

### CODE REVIEW

**Comment by Kyep** [comment]
This:

- Effectively turns all wizard rounds into ragin' mages.
- Removes 4 commonly-used wizard spells (slaughter demon, laughter demon, summon guns, summon magic)
- Starts off with "refactor" but is really a complete re-work of a round type that effectively turns it into a different round-type. That's arguably beyond the scope of something called a "refactor". "Refactor" generally means "optimizing the code without changing the functionality in a significant way". This is dramatic changes to functionality.

Ultimately, my issue with this is that if we wanted to have ragin' mages instead of wizard, we'd just have ragin' mages in rotation, not wizard. The reason they're different game mode types, and wizard is in rotation but ragin' mages isn't, is that we ultimately don't want ragin' mages to be a regular thing, because its among the highest chaos type of rounds on the server. It is okay for an occasional adminbus thing but if it replaced all wizard rounds it would make the server notably more chaotic.

I do like the fix to 8278, but the rest of it seems to effectively turn wizard into a clone of ragin' mages, which I don't support.

**Comment by datlo** [comment]
Raging mages has 1 wizard per 6 players and can have multiple wizards spawned at the same time. This has 1 wiz per 25. The PR extends a wizard round unlike raging mages which intensifies it. They are definitely different.

**Comment by Mitchs98** [comment]
I whole-heartedly agree with removing demons. They're generally unfun for anyone but the demon and sadistic wizard watching everyone die. They're an overly effective permakill that's nearly impossible to kill at any given time. Sure, it fits the niche of 'chaos'. But, most people I've seen just dislike them.

Not to mention in my opinion they're relatively boring and are nothing but 'Camp medbay and eat everyone hurrdurr'. Wizard should have to work harder for their goals rather than summon a demon and hide in a locker all round.

That said. I disagree with removing summon guns or summon magic. Not really a fan of making every round mini-ragin' either given the wizards' permakill/nigh permakill spells(I.E cluwne/gib). If cluwne was reworked, it'd be fine...maybe. Then again there are those wizards that go around gibbing everyone and destroying their brain. So. Eh.

**Comment by Fox-McCloud** [comment]
Major 👎 to removing slaughter or laughter demons. I will refuse to support this PR in any form if that's included.

I'd much rather summon guns be changed to make a portion of the crew antagonistic towards each other than remove it.

**Comment by datlo** [comment]
I suppose the demons can stay. Removing them is a little out of scope anyway.

For summon guns and magic Ive thought of a solution to keep the spells. I'll delete the items once the wizard whos summoned them died, that way later wizards are not affected by the summon. I'll get to implementing that.

**Comment by SkeletalElite** [comment]
> I suppose the demons can stay. Removing them is a little out of scope anyway.
> 
> For summon guns and magic Ive thought of a solution to keep the spells. I'll delete the items once the wizard whos summoned them died, that way later wizards are not affected by the summon. I'll get to implementing that.

You could also go with the TG route where summon guns and magic gives part of the crew survival antag objectives so that the crew startings killing each other with the weapons and magic.

**Comment by Ralta** [comment]
Nice. Wizard rounds are rarely fun, as it stands. 5-15 minutes of a wizard doing wizard-things which is fun for only a handful of people.

As suggested Skeletal, it'd be nice if a few crew became mini-antags as part and parcel of certain spells being used in order to keep things interesting.

**Comment by taukausanake** [comment]
I rather just have raging mages in rotation and single wizards as a random antag event. This seems like it might work, so long as the removed spells are returned. Kind of weird that the SWF would send another, _single_ wizard to the station that just killed the previous one. Why I still feel raging mages would be the better game round; think nuke ops wizard team minus the nuke

**Comment by MrMagolor** [comment]
I think just having multiple wizards spawn in roundstart depending on pop, and possibly reducing the total spell points for wiz, so there can be more wizards, would reduce the "wizard once a year" memes, as well as making the rounds more interesting.

_This is getting out of hand. Now there are two of them!_

**Comment by MrMagolor** [comment]
> You could also go with the TG route where summon guns and magic gives part of the crew survival antag objectives so that the crew startings killing each other with the weapons and magic.

Survival antag objectives?

**Comment by Emanthealmighty** [comment]
Perhaps there could be a points system that allows wizards to grow stronger and get better powers as they live, so we could also technically keep demons because they would be way up there with their cost.

The starting wizard would have the biggest starting amount of points, and as the wizards die, the next ones would gradually have less and less, forcing them to take different and less chaotic approaches every time, giving the crew members time to regain their composure (if they even killed the starting wizard). The time between the wizards spawning should also increase as they die, further helping the crew.

**Comment by datlo** [comment]
Big update. Demons are back without change, summon guns and magic are back but dissapear when a new wizard spawn, and lich was reworked. The rework was necessary because it was extremely difficult/spaghetti to tell when a lich was actually dead or not currently. Now its much more simple, if the lich isnt gibbed/dusted, it can eventually revive. Read the updated PR description for details.

File: code/modules/mob/living/death.dm [hunk 1]
```diff
@@ -88,6 +88,9 @@
 	if(mind)
 		mind.store_memory("Time of death: [station_time_timestamp("hh:mm:ss", timeofdeath)]", 0)
 		GLOB.respawnable_list += src
+		for(var/obj/item/I in mind.summoned_items)
+			I.loc.visible_message("<span class='warning'>[I] suddenly disappears!</span>")
```
Comment (Fox-McCloud) [inline code/modules/mob/living/death.dm]: if you're going to have the item have the `visible_message`, do that, otherwise, you should be using the turf.

File: code/modules/mob/living/death.dm [hunk 2]
```diff
@@ -88,6 +88,9 @@
 	if(mind)
 		mind.store_memory("Time of death: [station_time_timestamp("hh:mm:ss", timeofdeath)]", 0)
 		GLOB.respawnable_list += src
+		for(var/obj/item/I in mind.summoned_items)
```
Comment (Fox-McCloud) [inline code/modules/mob/living/death.dm]: This is never used.

File: code/datums/mind.dm [hunk 3]
```diff
@@ -34,6 +34,7 @@
 	var/list/restricted_roles = list()
 
 	var/list/spell_list = list() // Wizard mode & "Give Spell" badmin button.
+	var/list/summoned_items = list() //List of items summoned by this wizard. Deleted when a new wizard is spawned.
```
Comment (Fox-McCloud) [inline code/datums/mind.dm]: never used

File: code/game/gamemodes/wizard/wizard.dm [hunk 4]
```diff
@@ -165,38 +173,138 @@
...
+
+/datum/game_mode/wizard/proc/make_more_mages()
+	if(making_mage || SSshuttle.emergency.mode >= SHUTTLE_ESCAPE)
+		return FALSE
+	making_mage = TRUE
+	var/list/candidates = list()
+	var/mob/dead/observer/harry = null
+	spawn(rand(200, 600))
+		message_admins("SWF is still pissed, sending another wizard.")
+		for(var/obj/item/I in summoned_items) //Wipe all summoned items so the new wiz isnt fucked over by a previous one.
```
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/wizard.dm]: Not removing it from the list means it will never garbage collect.

Check your gosh darn references.

File: code/game/gamemodes/wizard/wizard.dm [hunk 5]
```diff
@@ -165,38 +173,138 @@
...
+		// No goodies for you
+		qdel(B)
+
+/datum/game_mode/wizard/proc/make_more_mages()
+	if(making_mage || SSshuttle.emergency.mode >= SHUTTLE_ESCAPE)
+		return FALSE
+	making_mage = TRUE
+	var/list/candidates = list()
+	var/mob/dead/observer/harry = null
+	spawn(rand(200, 600))
```
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/wizard.dm]: `addtimer`. 

Stop using spawns.

File: code/datums/spells/lichdom.dm [hunk 6]
```diff
@@ -89,45 +80,89 @@
...
 					H.unEquip(H.head)
 					H.equip_to_slot_or_del(new /obj/item/clothing/suit/wizrobe/black(H), slot_wear_suit)
 					H.equip_to_slot_or_del(new /obj/item/clothing/head/wizard/black(H), slot_head)
 					H.equip_to_slot_or_del(new /obj/item/clothing/shoes/sandal(H), slot_shoes)
-					H.equip_to_slot_or_del(new /obj/item/clothing/under/color/black(H), slot_w_uniform)
\ No newline at end of file
+					H.equip_to_slot_or_del(new /obj/item/clothing/under/color/black(H), slot_w_uniform)
+
+
+/obj/item/phylactery
```
Comment (Fox-McCloud) [inline code/datums/spells/lichdom.dm]: I don't really like this, to be honest.

Ensouled Rubber Duckies, toilet water filled beakers, plushies, and what have you are great and allows the lich to use creative things.

Comment (Mitchs98) [reply code/datums/spells/lichdom.dm]: Honestly agree. It allows creativity and makes the crew have to work harder. Which they should have to do, rather than just looking for a green heart somewhere. Sure you can probably hide it well..but still.

Comment (datlo) [reply code/datums/spells/lichdom.dm]: I will look into a way to have this work with any item.

**Comment by datlo** [comment]
Unfortunately I don't see a clean way to keep custom items as a phylactery. Either it's purely cosmetic and the items don't actually keep their properties, which is confusing and does not provide much gameplay value, or it's not and we run into the issues of using a light bulb in a socket and other exploit-y thing.

The phylactery is a small item and there are plenty of creative ways you can already hide a small item on station.

Unless somebody has a better idea on how to implement this, I am keeping the lich rework design as it is for now.

**Comment by TDSSS** [comment]
IMO, just making the items look like they did before would already be an improvement, making it much easier to hide it than having it look like a certain, very magical thing.

**Comment by MrMagolor** [comment]
> The phylactery is a small item and there are plenty of creative ways you can already hide a small item on station.

Except... you wrote that the phylactery will instantly snap back to the lich, so you'd have to re-hide it _every_ time.

**Comment by datlo** [comment]
Closing this for now. I will probably come back to this, but anyone who wants to use the code feel free to.

**Comment by datlo** [comment]
Reopening this since I figured out a clean implementation. Liches still work the same as before with any item you want, but you dust if the item is destroyed and it teleports to you if youre on separate zlevels. 

This doesnt fix the exploit-y issues... but thatll be for another PR.

**Comment by datlo** [comment]
I am a little scared to say so but I beliiieve this is ready for merge?

If approved this should probably be testmerged because theres a fuckton of possible edge cases and I cant test absolutely everything by myself. Also need to see if the gamemode flows well in an actual live environment.

**Comment by datlo** [comment]
Hahahahahaha no of course itd be too easy. Still WIP, bugs to fix.

**Comment by datlo** [comment]
Wont be finishing the PR at least for a while. There wouldnt be much work needed to wrap it up so if anyone wants to pick it up feel free.
